### ISSUE TITLE
Bad attention weights

### ISSUE THREAD

**Issue by dLeos** [NVIDIA/flowtron#43]
Hello

I am trying to train Flowtron on LJSpeech
Unfortunately after 24 hours of training, the attention weights still bad
Server configuration: 4 instances with 8xV100

![image](https://user-images.githubusercontent.com/15457550/87042212-8df97f00-c21d-11ea-824b-07c32200571f.png)

![image](https://user-images.githubusercontent.com/15457550/87043159-d7969980-c21e-11ea-9899-3c8c6e4c3026.png)

Do you have any ideas?

**Comment by rafaelvalle** [comment]
Yes, train it by progressively adding steps of flow as the model learns to attend on each step of flow.
Start with 1 step of flow, train it until it learns attention, use this model to warm-start a model with 2 steps of flow, and so on...

**Comment by dLeos** [comment]
@rafaelvalle Thanks for your reply

I ran new training with option model_config.n_flows=1, but after 16 hours attention weights look bad again:

![image](https://user-images.githubusercontent.com/15457550/87129193-22fe8580-c2bb-11ea-97df-6a4c0116042c.png)

In one of the threads I read that good alignment is produced in less than 24 hours.

So, what could be wrong?

**Comment by rafaelvalle** [comment]
Can you share your tensorboard plots?

**Comment by dLeos** [comment]
Yes

![image](https://user-images.githubusercontent.com/15457550/87170989-47c71d00-c2fc-11ea-8c5b-94df1685f6af.png)

![image](https://user-images.githubusercontent.com/15457550/87170886-26663100-c2fc-11ea-9ba8-eee408d024d4.png)

**Comment by rafaelvalle** [comment]
Does it have good attention around 60k iters
?

**Comment by dLeos** [comment]
Not. Attention on all iterations looks the same

**Comment by rafaelvalle** [comment]
Make sure you trim silences from the beginning and end of your audio files

**Comment by dLeos** [comment]
I use LJSpeech dataset for training. Any instructions on how to trim them?

Could the problem be that I use distributed training?

Also, I set the flag fp16_run=true

**Comment by adrianastan** [comment]
> Make sure you trim silences from the beginning and end of your audio files

Should there be no silence at all in the beginning and end, or should there be at least, let's say 0.1 seconds of silence?

**Comment by adrianastan** [comment]
> I use LJSpeech dataset for training. Any instructions on how to trim them?

The simplest way would be to use [librosa.effect.trim()](https://librosa.org/librosa/generated/librosa.effects.trim.html?highlight=librosa%20effects%20trim)

**Comment by rafaelvalle** [comment]
there should be no silence at all.
sox and librosa.effects.trim can be used to trim silences from beginning and end

**Comment by kurbobo** [comment]
I have similar problem

> I use LJSpeech dataset for training. Any instructions on how to trim them?
> 
> Could the problem be that I use distributed training?
> 
> Also, I set the flag fp16_run=true

Have you solved this problem?
Also I tried to predict not mel spectrogram but lpc spectrogram but always got such picture, does anybody know what is the problem?
![image](https://user-images.githubusercontent.com/32241519/87778245-92ccbd00-c833-11ea-8dee-6c23d42b6938.png)

**Comment by dLeos** [comment]
The problem remained unresolved.
I tried to trim silences from the beginning and end of audio files with the librosa.effects.trim(), but the picture remains the same.

**Comment by rafaelvalle** [comment]
@kurbob does the attention map always look like that? You might have to change from byte to bool
https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

**Comment by rafaelvalle** [comment]
@adrianastan There should be **no silence** at the beginning or at the end of an audio file.

**Comment by zjFFFFFF** [comment]
@rafaelvalle  Can you tell me what the meaning of no silence is?  If i use librosa.effects.trim(),  top_db should be set to what. 
For my data set, if I set top_db to 20, some sounds will also be cut off. setting it a little higher, it seems that some audio files still have silence at the beginning.

**Comment by dLeos** [comment]
@kurbobo The problem was solved when I used encoder and embedding layers from pretrained model.

@zjFFFFFF In my case top_db = 30 works well enough

**Comment by zjFFFFFF** [comment]
@dLeos

In fact, I got the same plot as you(training from scratch). But the validation loss does not seem to affect the experimental results. Between iterations: 800,000-950,000, the model can generate acceptable sounds. So you can try different checkpoints one by one.

**Comment by zjFFFFFF** [comment]
@dLeos

In fact, I got the same plot as you(training from scratch). But the validation loss does not seem to affect the experimental results. Between iterations: 800,000-950,000, the model can generate acceptable sounds. So you can try different checkpoints one by one.

**Comment by zjFFFFFF** [comment]
@dLeos

In fact, I got the same plot as you(training from scratch). But the validation loss does not seem to affect the experimental results. Between iterations: 800,000-950,000, the model can generate acceptable sounds. So you can try different checkpoints one by one.

**Comment by kurbobo** [comment]
> @kurbob does the attention map always look like that? You might have to change from byte to bool
> https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33
@rafaelvalle 
No, it's not always appears and I had already fixed the "type/bool" problem before train, but nevertheless sometimes such problem happens.
I have one more question: am I right, that flowtron in this repo converts every sentence to arpabet transciption and then train to map sequence of arpabet transcriptions to sqauence of frequency frames?

**Comment by kurbobo** [comment]
> > @kurbob does the attention map always look like that? You might have to change from byte to bool
> > https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33
> > @rafaelvalle
> > No, it's not always appears and I had already fixed the "type/bool" problem before train, but nevertheless sometimes such problem happens.
> > I have one more question: am I right, that flowtron in this repo converts every sentence to arpabet transciption and then train to map sequence of arpabet transcriptions to sqauence of frequency frames?

@rafaelvalle

**Comment by Liujingxiu23** [comment]
@kurbobo @rafaelvalle  I tried mels , train n-flows =1 first and then use the model to warmup  n-flows = 2 model , the two alignment weights are both right, and the wavs synthsized are good.  But to lpc parameter that used in lpcnet vocoder, when n-flows=1, everythings seems good,  loss is good, alignment is right,  however when I train n-flows=2 with the trained n-flows=1 model as warmup, the second alignment failed, and the loss just vibrate without any descend.

**Comment by rafaelvalle** [comment]
@Liujingxiu23 please share training, validations losses and attention for 1 step of flow model and 2 steps of flow model.

**Comment by Liujingxiu23** [comment]
@rafaelvalle  
when using mels, the mels are changed to db and normed to [-4,4] the way used in tacotron tf-version, both n-flow=1 or 2, models are good.
When using lpcs that used in lpcnet vocoder, the parameters are also norm to [-4,4] using min-max norm
n-flow=1:
![lpc-nflow=1-loss](https://user-images.githubusercontent.com/49260794/93570133-2140f480-f9c5-11ea-991c-8b66ffaee59e.PNG)
![lpc-nflow=1-attention](https://user-images.githubusercontent.com/49260794/93570142-24d47b80-f9c5-11ea-9a7f-bdd81c13ff42.PNG)
n-flow=2 warmup with n-flow=1 model:
![lpc-nflow=2-loss](https://user-images.githubusercontent.com/49260794/93570173-3027a700-f9c5-11ea-94cd-76fda93a093e.PNG)
![lpc-nflow=2-attention](https://user-images.githubusercontent.com/49260794/93570182-328a0100-f9c5-11ea-8b36-fb566d4b69f4.PNG)

**Comment by rafaelvalle** [comment]
Did you warmup the 2 flows model with the 1 flow model from a checkpoint around 200k?

**Comment by Liujingxiu23** [comment]
@rafaelvalle  Thank you for you reply!
I did not warmup from 200k, but about 500k. 
I retrain the n-flow=2 model warmup from 200k, the loss decline with shake.

![lpc-nflow=2-loss-warmup-from-200k](https://user-images.githubusercontent.com/49260794/93741835-b2aba300-fc1f-11ea-89d8-cbd4e17ecb4d.PNG)

Compared to mels, lpc parameter seems harder to train. the wavs synthesized by lpcnet with n-flow=1 is worse than that of mels under the same situation.
I donot know whether I should use the origial lpc parameter without norm.
